Add support to submit bio request which are processed synchoronously.
Melio requires this runtime behaviour in order to work correctly.
diff --git a/block/blk-core.c b/block/blk-core.c
index 5a750b1..3049714 100644
--- a/block/blk-core.c
+++ b/block/blk-core.c
@@ -1803,6 +1803,13 @@ void generic_make_request(struct bio *bio)
 	if (!generic_make_request_checks(bio))
 		return;
 
+	if (current->bio_sync_dispatch)
+	{
+		struct request_queue *q = bdev_get_queue(bio->bi_bdev);
+		q->make_request_fn(q, bio);
+		return;
+	}
+
 	/*
 	 * We only want one ->make_request_fn to be active at a time, else
 	 * stack usage with stacked devices could be a problem.  So use
@@ -1847,6 +1854,23 @@ void generic_make_request(struct bio *bio)
 EXPORT_SYMBOL(generic_make_request);
 
 /**
+ * generic_make_sync_request - like generic_make_request, but dispatches immediately
+ * @bio:  The bio describing the location in memory and on the device.
+ *
+ * generic_make_sync_request() can be used in modules that wait on the completion
+ * of the bio immediately after this call.
+ */
+void generic_make_sync_request(struct bio *bio)
+{
+	int old_bio_dispatch = current->bio_sync_dispatch;
+
+	current->bio_sync_dispatch = 1;
+	generic_make_request(bio);
+	current->bio_sync_dispatch = old_bio_dispatch;
+}
+EXPORT_SYMBOL(generic_make_sync_request);
+
+/**
  * submit_bio - submit a bio to the block device layer for I/O
  * @rw: whether to %READ or %WRITE, or maybe to %READA (read ahead)
  * @bio: The &struct bio which describes the I/O
diff --git a/include/linux/blkdev.h b/include/linux/blkdev.h
index 494d228..44bfe34 100644
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -731,6 +731,7 @@ static inline void rq_flush_dcache_pages(struct request *rq)
 extern int blk_register_queue(struct gendisk *disk);
 extern void blk_unregister_queue(struct gendisk *disk);
 extern void generic_make_request(struct bio *bio);
+extern void generic_make_sync_request(struct bio *bio);
 extern void blk_rq_init(struct request_queue *q, struct request *rq);
 extern void blk_put_request(struct request *);
 extern void __blk_put_request(struct request_queue *, struct request *);
diff --git a/include/linux/sched.h b/include/linux/sched.h
index 00c1d4f..93f03c8 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1288,6 +1288,7 @@ struct task_struct {
 	void *journal_info;
 
 /* stacked block device info */
+	int bio_sync_dispatch;
 	struct bio_list *bio_list;
 
 #ifdef CONFIG_BLOCK
